File: vane-core/src/main/java/org/oddlama/vane/util/ResourceList.java
package org.oddlama.vane.util;

import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.regex.Pattern;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

/**
 * from: forums.devx.com/showthread.php?t=153784 list resources available from the jar file of the
 * given class
 */
public class ResourceList {

    /**
     * For all elements of java.class.path get a Collection of resource Pattern pattern =
     * Pattern.compile(".*"); gets all resources
     *
     * @param pattern the pattern to match
     * @return the resources in the order they are found
     */
    public static Collection<String> getResources(final Class<?> clazz, final Pattern pattern) {
        final var jarUrl = clazz.getProtectionDomain().getCodeSource().getLocation();
        try {
            return getResources(new URI(jarUrl.toString()).getPath(), pattern);
        } catch (URISyntaxException e) {
            return new ArrayList<String>();
        }
    }

    private static Collection<String> getResources(final String path, final Pattern pattern) {
        final var retval = new ArrayList<String>();
        final var file = new File(path);
        if (file.isDirectory()) {
            retval.addAll(getResourcesFromDirectory(file, pattern));
        } else {
            retval.addAll(getResourcesFromJarFile(file, pattern));
        }
        return retval;
    }

    private static Collection<String> getResourcesFromJarFile(final File file, final Pattern pattern) {
        final var retval = new ArrayList<String>();
        ZipFile zf;
        try {
            zf = new ZipFile(file);
        } catch (final IOException e) {
            throw new Error(e);
        }
        final var e = zf.entries();
        while (e.hasMoreElements()) {
            final ZipEntry ze = e.nextElement();
            final String fileName = ze.getName();
            final boolean accept = pattern.matcher(fileName).matches();
            if (accept) {
                retval.add(fileName);
            }
        }
        try {
            zf.close();
        } catch (final IOException e1) {
            throw new Error(e1);
        }
        return retval;
    }

    private static Collection<String> getResourcesFromDirectory(final File directory, final Pattern pattern) {
        final var retval = new ArrayList<String>();
        final var fileList = directory.listFiles();
        for (final File file : fileList) {
            if (file.isDirectory()) {
                retval.addAll(getResourcesFromDirectory(file, pattern));
            } else {
                try {
                    final String fileName = file.getCanonicalPath();
                    final boolean accept = pattern.matcher(fileName).matches();
                    if (accept) {
                        retval.add(fileName);
                    }
                } catch (final IOException e) {
                    throw new Error(e);
                }
            }
        }
        return retval;
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/util/ResourceList.kt
package org.oddlama.vane.util

import java.io.IOException
import java.net.URI
import java.net.URISyntaxException
import java.util.regex.Pattern
import java.util.zip.ZipFile

/**
 * Lists resources available from the code source of a given class.
 */
object ResourceList {
    /**
     * For all elements of java.class.path get a Collection of resources matching [pattern].
     *
     * @param pattern the pattern to match; use `Pattern.compile(".*")` to get all resources
     * @return the resources in the order they are found
     */
    @JvmStatic
    fun getResources(clazz: Class<*>, pattern: Pattern): List<String> {
        val jarUrl = clazz.protectionDomain.codeSource.location
        return try {
            getResources(URI(jarUrl.toString()).path, pattern)
        } catch (_: URISyntaxException) {
            emptyList()
        }
    }

    /** Resolves resources from either a directory or jar path. */
    private fun getResources(path: String, pattern: Pattern): List<String> {
        val file = java.io.File(path)
        return when {
            file.isDirectory -> getResourcesFromDirectory(file, pattern)
            else -> getResourcesFromJarFile(file, pattern)
        }
    }

    /** Collects matching resource names from a jar file. */
    private fun getResourcesFromJarFile(file: java.io.File, pattern: Pattern): List<String> {
        return try {
            ZipFile(file).use { zf ->
                zf.entries().asSequence()
                    .map { it.name }
                    .filter { pattern.matcher(it).matches() }
                    .toList()
            }
        } catch (e: IOException) {
            throw Error(e)
        }
    }

    /** Recursively collects matching resource paths from a directory tree. */
    private fun getResourcesFromDirectory(directory: java.io.File, pattern: Pattern): List<String> {
        return buildList {
            directory.listFiles()?.forEach { file ->
                when {
                    file.isDirectory -> addAll(getResourcesFromDirectory(file, pattern))
                    else -> try {
                        val fileName = file.canonicalPath
                        if (pattern.matcher(fileName).matches()) add(fileName)
                    } catch (e: IOException) {
                        throw Error(e)
                    }
                }
            }
        }
    }
}
